dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,352 @@
+
+import ops
+from charms.bind.v0 import dns_record
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseEndpointsChangedEvent,
+)
+from charms.operator_libs_linux.v1 import systemd
+
+import constants
+import models
+import templates
+from database import DatabaseHandler
+from dns_policy import DnsPolicyService
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+class ReconciliateEvent(ops.charm.EventBase):
+    """Event representing a periodic reload of the charmed-bind service."""
+
+
+class DnsPolicyCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+
+        self.on.define_event("reconciliate", ReconciliateEvent)
+        logger.debug("unit name: %s", self.unit.name)
+
+        self.dns_policy = DnsPolicyService()
+        self._database = DatabaseHandler(self, constants.DATABASE_RELATION_NAME)
+        self.dns_record_provider = dns_record.DNSRecordProvides(self, "dns-record-provider")
+        self.dns_record_requirer = dns_record.DNSRecordRequires(self, "dns-record-requirer")
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self._database.database.on.database_created, self._on_database_created
+        )
+        self.framework.observe(
+            self._database.database.on.endpoints_changed, self._on_database_endpoints_changed
+        )
+        self.framework.observe(
+            self.on[constants.DATABASE_RELATION_NAME].relation_broken,
+            self._on_database_relation_broken,
+        )
+        self.framework.observe(self.on.create_reviewer_action, self._on_create_reviewer_action)
+        self.framework.observe(
+            self.dns_record_provider.on.dns_record_request_received,
+            self._on_dns_record_request_received,
+        )
+        self.framework.observe(self.on.reconciliate, self._on_reconciliate)
+        self.unit.open_port("tcp", 8080)  # dns-policy-app
+
+    def _on_reconciliate(self, _: ReconciliateEvent) -> None:
+        relation_data = self._get_remote_relation_data()
+        if relation_data is None:
+            return
+        dns_policy_config = self.dns_policy.get_config()
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(entry)
+        # logger.debug("Entries: %s", entries)
+        if not entries:
+            return
+        logger.debug("Config: %s", dns_policy_config)
```

Comment:
If the debug log is still needed put it next to where the config is fetched

Reply:
It was removed